chia/server/ws_connection.py

```diff
@@ -169,6 +169,9 @@ def create(
             session=session,
         )
 
+    def __hash__(self) -> int:
+        return hash(self.peer_node_id)
```

Comment:
Are we really sure this is always unique at all times?  We've definitely had issues with mixing up old and reconnected connections.

Reply:
I'm trying to remember the details around this and collisions and checking equality and...  I'll have to think this over a bit later.

Reply:
https://docs.python.org/3.11/reference/datamodel.html#object.__hash__
> The only required property is that objects which compare equal have the same hash value

Also read the rest about mutable attributes etc.  I think we can do this if we really want to, but have to appropriately handle both `.__hash__()` and `.__eq__()` in compatible ways.  `dataclasses(unsafe_hash=)` also has commentary.

https://docs.python.org/3.11/library/dataclasses.html#dataclasses.dataclass

I'm guessing that we'll end up delegating to `object.__hash__()` and `object.__eq__()`, but think it over if we really want to make this mutable class hashable.

Reply:
> Are we really sure this is always unique at all times? We've definitely had issues with mixing up old and reconnected connections.

Yeah they are unique as they are calculated from the client certs. Sure those can be reused on different clients to force a misuse and duplication but this shouldn't affect using it as identifier. We already use the peer id as identifier for connections spread all over the codebase so i don't see an issue with doing this. 


> I think we can do this if we really want to, but have to appropriately handle both .__hash__() and .__eq__() in compatible ways.  dataclasses(unsafe_hash=) also has commentary.

What if we don't handle `__eq__`? I don't care about which way we do it so just let me know what you prefer. 

> I'm guessing that we'll end up delegating to object.__hash__() and object.__eq__(), but think it over if we really want to make this mutable class hashable.

Why not? Whats the downside of doing it? I feel like its nice to be able to just put connections into sets or dicts instead of putting their ids in there and then wherever you want to use them fetch them again from the server based on the id.

Reply:
> > Are we really sure this is always unique at all times? We've definitely had issues with mixing up old and reconnected connections.
> 
> Yeah they are unique as they are calculated from the client certs. Sure those can be reused on different clients to force a misuse and duplication but this shouldn't affect using it as identifier. We already use the peer id as identifier for connections spread all over the codebase so i don't see an issue with doing this.

What about when we initiate a new connection to a node we are already connected to?  We've already had bugs around this.  It is actually extremely common for this to happen on small testnets since you never have as many peers as you want even when connected to every peer on the network so you are always trying to connect to every single peer even though already connected to them.  Oh yeah, with IPv4 and IPv6...

> > I think we can do this if we really want to, but have to appropriately handle both .**hash**() and .**eq**() in compatible ways.  dataclasses(unsafe_hash=) also has commentary.
> 
> What if we don't handle `__eq__`? I don't care about which way we do it so just let me know what you prefer.

The `dict` and `set` etc assume that you follow the rules laid out for `.__hash__()` and `.__eq__()`.  I could look further into their exact failure mode, but I think we just want to avoid these basic things working incorrectly.  Due to hash collisions, equality checks are part of lookups.

> > I'm guessing that we'll end up delegating to object.**hash**() and object.**eq**(), but think it over if we really want to make this mutable class hashable.
> 
> Why not? Whats the downside of doing it? I feel like its nice to be able to just put connections into sets or dicts instead of putting their ids in there and then wherever you want to use them fetch them again from the server based on the id.

The downside is that we exercise the misassumptions and incorrect implementations already laid out here `:]` resulting in bugs lots of us wouldn't be aware are possible.  I'm not saying a hard no to this pr, but I am suggesting we are careful about both the intent and the implementation.

Note that the `object` methods are written around using identity (`is`) for equality checks so they kind of side step the issues.  Not good when you want to be able to check if two objects are, well, equal...  but useful when you want to consider every instance to be unique even with the same attributes.